### PR TITLE
Added cardinal semiotic class for te ITN

### PR DESCRIPTION
**Pull request by jaya-TN**
# What does this PR do ?

This PR adds support for the Telugu Inverse Text Normalization (ITN) Cardinal semiotic class.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Have you signed your commits? Use ``git commit -s`` to sign.
- [ ] Do all unittests finish successfully before sending PR?
   1) ``pytest`` or (if your machine does not have GPU) ``pytest --cpu`` from the root folder (given you marked your test cases accordingly `@pytest.mark.run_only_on('CPU')`).
   2) Sparrowhawk tests ``bash tools/text_processing_deployment/export_grammars.sh --MODE=test ...``
- [ ] If you are adding a new feature: Have you added test cases for both `pytest` and Sparrowhawk [here](tests/nemo_text_processing).
- [x] Have you added ``__init__.py`` for every folder and subfolder, including `data` folder which has .TSV files?
- [ ] Have you followed codeQL results and removed unused variables and imports (report is at the bottom of the PR in github review box) ?
- [x] Have you added the correct license header `Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.` to all newly added Python files?
- [x] If you copied [nemo_text_processing/text_normalization/en/graph_utils.py](nemo_text_processing/text_normalization/en/graph_utils.py) your header's second line should be `Copyright 2015 and onwards Google, Inc.`. See an example [here](https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/text_normalization/en/graph_utils.py#L2).
- [x] Remove import guards (`try import: ... except: ...`) if not already done.
- [] If you added a new language or a new feature please update the [NeMo documentation](https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/text_normalization/wfst/wfst_text_normalization.rst) (lives in different repo).
- [ ] Have you added your language support to [tools/text_processing_deployment/pynini_export.py](tools/text_processing_deployment/pynini_export.py).
  

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation
- [ ] Test

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/hundred.tsv
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/hundred.tsv:1]: why we need both hundred.tsv and hundred_words.tsv?

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/labels_exception.tsv
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/labels_exception.tsv:1]: blank files, please modify or delete it

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+        self.graph_single_digit_with_zero = pynutil.insert("౦") + graph_digit
+
+        self.graph_two_digit = (
+            graph_teens_and_ties
+            | graph_ties_plus_okkati
+            | (graph_ties_prefix + delete_space + graph_digit)
+            | self.graph_single_digit_with_zero
+        )
+        hundred_words = [row[0] for row in load_labels(get_abs_path("data/numbers/hundred_words.tsv"))]
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: why only row[0]?

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+
+MIN_NEG_WEIGHT = -0.0001
+MIN_POS_WEIGHT = 0.0001
+INPUT_CASED = "cased"
+INPUT_LOWER_CASED = "lower_cased"
+MINUS = pynini.union("ఋణాత్మక", "మైనస్").optimize()
+
+
+def integer_to_telugu(n: int) -> str:
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: why we need a customize function for join string?

**Review by folivoramanh — CHANGES_REQUESTED** [review]
- add sparrowhawk file and run sh test (we need to make sure both pytest and sparrowhawk pass)
- add te to jenkins file
- try to optimize cardinal logic (you can look up other languages eg. Hindi, Spanish, Portuguese, etc... for both TN and ITN) to optimize graph logic, too much looping could lead to slow runtime
- do te ITN support english number (eg. 123456789?) if yes, can you add test cases for both te number and en number?

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/labels_exception.tsv
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/labels_exception.tsv:1]: deleted

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 1 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: removed (magnitudes load via standard string_file).

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/hundred.tsv
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/hundred.tsv:1]: consolidated to one file per magnitude; hundred_words.tsv removed.

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py (hunk 2 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: custom join function removed.

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/crore.tsv
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/crore.tsv:1]: why first and second line got only 1 value while third row got 2?

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    Args:
+        input_case: accepting either "lower_cased" or "cased" input.
+    """
+
+    def __init__(self, input_case: str = INPUT_LOWER_CASED):
+        super().__init__(name="cardinal", kind="classify")
+        self.input_case = input_case
+
+        graph_zero = pynini.string_file(get_abs_path("data/numbers/zero.tsv")).invert()
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: do not abuse .invert(), try to standardize the data

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        graph_digit = pynini.string_file(get_abs_path("data/numbers/digit.tsv")).invert()
+        graph_teens_and_ties = pynini.string_file(get_abs_path("data/numbers/teens_and_ties.tsv")).invert()
+        graph_ties_prefix = pynini.string_file(get_abs_path("data/numbers/ties_prefix.tsv")).invert()
+
+        graph_people = pynini.string_file(get_abs_path("data/numbers/digit_people.tsv")).invert()
+        graph_special = pynini.string_file(get_abs_path("data/numbers/special_numbers.tsv")).invert()
+        two_te_digits = NEMO_TE_DIGIT + NEMO_TE_DIGIT
+        graph_digit |= (graph_people | graph_special) @ NEMO_TE_DIGIT
+
+        self.graph_zero = graph_zero
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: only set self.variable when it's reusable on other semiotic class

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        graph_two_digit = (
+            graph_teens_and_ties
+            | (graph_ties_prefix + delete_space + graph_digit)
+            | (pynutil.insert("౦") + graph_digit)
+            | ((graph_people | graph_special) @ two_te_digits)
+        )
+        self.graph_two_digit = graph_two_digit
+        two_digit_or_zeros = graph_two_digit | pynutil.insert("౦౦")
+
+        graph_case_suffix = pynini.string_file(get_abs_path("data/numbers/case_suffix.tsv"))
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: move all pynini.string_file call to begin of init function for easier checking file call

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: change all 2 first lines to 
"# Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
#"

File: tests/nemo_text_processing/te/test_cardinal.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+class TestCardinal:
+    tagger = ClassifyFst(overwrite_cache=True).fst
+    verbalizer = VerbalizeFinalFst().fst
+
+    @parameterized.expand(parse_test_case_file('te/data_inverse_text_normalization/test_cases_cardinal.txt'))
+    @pytest.mark.run_only_on('CPU')
+    @pytest.mark.unit
+    def test_denorm(self, test_input, expected):
+        tagged = pynini.shortestpath(test_input @ self.tagger).string()
```
Comment (folivoramanh) [inline tests/nemo_text_processing/te/test_cardinal.py]: read hindi test and use the same code, do not use pynini.shortestpath here

File: tools/text_processing_deployment/pynini_export.py [hunk 8]
```diff
@@ -140,7 +141,7 @@ def parse_args():
 if __name__ == '__main__':
     args = parse_args()
 
-    if args.language in ['ru', 'es_en', 'hi_en', 'mr'] and args.grammars == 'tn_grammars':
+    if args.language in ['ru', 'es_en', 'hi_en', 'mr', 'te'] and args.grammars == 'tn_grammars':
```
Comment (folivoramanh) [inline tools/text_processing_deployment/pynini_export.py]: we also have PIC for te TN, right? if yes, please remove this change

**Review by folivoramanh — CHANGES_REQUESTED** [review]

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/crore.tsv
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/crore.tsv:1]: `crore.tsv` is one `value<TAB>word` file: the plural forms (`కోట్ల`, `కోట్లు`) have an empty value so they're deleted when a count precedes them, while the standalone `కోటి` maps to `౦౦౦౦౦౦౧` (one crore) — split in code via `@ accep("")` for the delete forms and `@ closure(digit,1)` for the value.

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 3 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: standardized all number TSVs to spoken<TAB>written (word-first, like en/es/pt/fr) and removed every .invert() from cardinal.py.

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 4 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: removed all self. sub-graph attributes

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 5 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: grouped all pynini.string_file calls at the top of __init__ so every data file reference is in one place.

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 6 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: changed

File: tests/nemo_text_processing/te/test_cardinal.py (hunk 7 above)
Comment (jaya-TN) [reply tests/nemo_text_processing/te/test_cardinal.py]: changed now same as hi.

File: tools/text_processing_deployment/pynini_export.py (hunk 8 above)
Comment (jaya-TN) [reply tools/text_processing_deployment/pynini_export.py]: Reverted back 'that line te' is no longer in that Sparrowhawk TN-block list

**Comment by jaya-TN** [comment]
Hi @folivoramanh, I've addressed all the requested changes, updated the PR, and requested a re-review.

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+import pynini
+from pynini import Far
+from pynini.export import export
+from pynini.lib import pynutil, utf8
+
+from nemo_text_processing.inverse_text_normalization.te.utils import load_labels
+
+NEMO_CHAR = utf8.VALID_UTF8_CHAR
+
+NEMO_TE_DIGIT = pynini.union("౦", "౧", "౨", "౩", "౪", "౫", "౬", "౭", "౮", "౯").optimize()
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/graph_utils.py:29]: is it digit.tsv?

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+    Args:
+        fst: input fst
+
+    Returns output fst where breaking spaces are converted to non breaking spaces
+    """
+    return fst @ pynini.cdrewrite(pynini.cross(NEMO_SPACE, NEMO_NON_BREAKING_SPACE), "", "", NEMO_SIGMA)
+
+
+def string_map_cased(input_file: str, input_case: str = INPUT_LOWER_CASED):
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: why we need it?

File: nemo_text_processing/inverse_text_normalization/te/utils.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    label_tsv = open(abs_path, encoding="utf-8")
+    labels = list(csv.reader(label_tsv, delimiter="\t"))
+    return labels
+
+
+from pynini.lib import pynutil
+
+
+def apply_fst(text, fst):
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/utils.py]: remove apply_fst, dont use it

**Review by folivoramanh — CHANGES_REQUESTED** [review]

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py (hunk 9 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/graph_utils.py:29]: digit.tsv maps spoken Telugu numbers to digits, while NEMO_TE_DIGIT is used as an acceptor for Telugu digit characters (0–9) in the FST graphs.

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py (hunk 10 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: We don't need string_map_cased for Telugu because Telugu doesn't have upper/lower case handling. It is also not used anywhere in the Telugu grammar, so I removed.

File: nemo_text_processing/inverse_text_normalization/te/utils.py (hunk 11 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/utils.py]: Removed apply_fst and the related unused import.

**Comment by jaya-TN** [comment]
Hi @folivoramanh, I've addressed all the requested changes, updated the PR.

**Comment by jaya-TN** [comment]
@folivoramanh folivoramanh
Hii,my pr is closed due to inactivity can u please reopen it again.

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/case_suffix.tsv
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/case_suffix.tsv:1]: Shipped but never referenced by any .py; the only orphan lexicon in the package. Contents are Telugu case suffixes, so this looks like intended handling that never landed. Wire it in or drop it.

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv [hunk 12]
```diff
@@ -0,0 +1,9 @@
+ఇద్దరు	౨
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv:2]: Series is complete 2–10 but omits 1. Verified: ఒక్కరు → ఒక్కరు (unchanged, falls through to word). is it a bug or relate to linguistic aspect?

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+ముగ్గురు	౩
+నలుగురు	౪
+ఐదుగురు	౫
+ఆరుగురు	౬
+ఏడుగురు	౭
+ఎనిమిదుగురు	౮
+తొమ్మిదుగురు	౯
+పది మంది	౧౦
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv]: he పది మంది → ౧౦ row deletes the classifier మంది, which every other count preserves:

ఆరు మంది ఆటగాళ్లు  →  ౬ మంది ఆటగాళ్లు    (kept — this is their own test case 2)
పది మంది ఆటగాళ్లు  →  ౧౦ ఆటగాళ్లు        (dropped)

suggestion: delete the row (పది already maps via teens_and_ties.tsv) → ౧౦ మంది ఆటగాళ్లు, 51/51 still pass.

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
+# Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/graph_utils.py:3]: missing the Google attribution line.

File: tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+నూట రెండు కోట్ల ఇరవై మూడు లక్షల నలభై ఐదు వేల ఐదు వందల అరవై ఏడు~౧౦౨౨౩౪౫౫౬౭
+వెయ్యి నూట రెండు కోట్ల ఇరవై మూడు లక్షల నలభై ఐదు వేల ఐదు వందల అరవై ఏడు~౧౧౦౨౨౩౪౫౫౬౭
+ఐదు వేల నూట రెండు కోట్ల ఇరవై మూడు లక్షల నలభై ఐదు వేల ఐదు వందల అరవై ఏడు~౫౧౦౨౨౩౪౫౫౬౭
+సున్నా ఐదు~౦౫
+సున్నా ఒకటి~౦౧
+సున్నా ఏడు మూడు~౦౭౩
+సున్నా సున్నా సున్నా ఒకటి~౦౦౦౧
+సున్నా సున్నా సున్నా~౦౦౦
+మైనస్ ఇరవై మూడు~-౨౩
+ఋణాత్మక ఐదు~-౫
```
Comment (folivoramanh) [inline tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt:51]: 15 derived lines to append: lakh/crore standalone at 1×/2×, the entirely-untested digit_people lexicon, the second పందొమ్మిది spelling of 19, and the 14-digit ceiling. Every expected value copied from an actual run.

**Review by folivoramanh — CHANGES_REQUESTED** [review]
minor changes, other LGTM

File: tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt (hunk 15 above)
Comment (jaya-TN) [reply tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt:51]: Added the 15 requested test cases covering standalone 1×/2× lakh and crore, all digit_people entries, the alternate పందొమ్మిది spelling, and the 14-digit upper-bound case. Verified the expected outputs against the current grammar. All Cardinal unit tests pass .

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py (hunk 14 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/graph_utils.py:3]: Added.

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv (hunk 13 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv]: Removed the పది మంది → ౧౦ mapping from digit_people.tsv so the classifier మంది is preserved. Updated the corresponding test expectation to పది మంది → ౧౦ మంది.

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv (hunk 12 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/digit_people.tsv:2]: This was a linguistic coverage gap. Added the singular human-count form ఒక్కరు → ౧ to digit_people.tsv .

File: nemo_text_processing/inverse_text_normalization/te/data/numbers/case_suffix.tsv
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/data/numbers/case_suffix.tsv:1]: case_suffix.tsv was not referenced by any grammar and was not required by the current Cardinal implementation, so I removed the unused lexicon.

**Comment by jaya-TN** [comment]
Hi @folivoramanh , I've addressed all the requested changes, updated the PR.

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright 2024 and onwards Google, Inc.
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: change 2024 to 2015

**Review by folivoramanh — CHANGES_REQUESTED** [review]
last change

File: nemo_text_processing/inverse_text_normalization/te/graph_utils.py (hunk 16 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/graph_utils.py]: Done.

**Comment by jaya-TN** [comment]
Hi @folivoramanh  , I've addressed all the requested changes, updated the PR.

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+        graph_ties_prefix = pynini.string_file(get_abs_path("data/numbers/ties_prefix.tsv"))
+        graph_people = pynini.string_file(get_abs_path("data/numbers/digit_people.tsv"))
+        graph_special = pynini.string_file(get_abs_path("data/numbers/special_numbers.tsv"))
+        graph_hundred = pynini.string_file(get_abs_path("data/numbers/hundred.tsv"))
+        graph_thousand = pynini.string_file(get_abs_path("data/numbers/thousand.tsv"))
+        graph_lakh = pynini.string_file(get_abs_path("data/numbers/lakh.tsv"))
+        graph_crore = pynini.string_file(get_abs_path("data/numbers/crore.tsv"))
+
+        two_te_digits = NEMO_TE_DIGIT + NEMO_TE_DIGIT
+        graph_digit |= (graph_people | graph_special) @ NEMO_TE_DIGIT
```
Comment (folivoramanh) [inline nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: **optional** · linguistic-claim

`graph_digit |= (graph_people | graph_special) @ NEMO_TE_DIGIT` folds the person-counting series (`digit_people.tsv`) into every generic digit slot, so it's usable as a multiplier before `వేల`/`లక్షల`/`వందలు` too — verified: `ఇద్దరు వేల` → ౨౦౦౦ (2000), `ముగ్గురు లక్షలు` → ౩౦౦౦౦౦ (300000). Flagging only to confirm this is deliberate rather than incidental (`digit_people.tsv`'s own naming and its standalone test cases at L56-64 of `test_cases_cardinal.txt` suggest it's meant for counting people specifically) — if these are genuinely attested Telugu constructions, no change needed; if not, worth scoping `graph_people` to the standalone-cardinal position only.

File: tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+ముగ్గురు~౩
+నలుగురు~౪
+ఐదుగురు~౫
+ఆరుగురు~౬
+ఏడుగురు~౭
+ఎనిమిదుగురు~౮
+తొమ్మిదుగురు~౯
+పది మంది~౧౦ మంది
+పందొమ్మిది~౧౯
+తొంభై తొమ్మిది లక్షల తొంభై తొమ్మిది వేల తొమ్మిది వందల తొంభై తొమ్మిది కోట్ల తొంభై తొమ్మిది లక్షల తొంభై తొమ్మిది వేల తొమ్మిది వందల తొంభై తొమ్మిది~౯౯౯౯౯౯౯౯౯౯౯౯౯౯
```
Comment (folivoramanh) [inline tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt:67]: **optional** · derived

No test case pins down the intended behavior of `digit_people` combined with a magnitude word (`వందలు`/`వేల`/`లక్షల`) either way. Once the question on `cardinal.py:56` is resolved, consider adding a case that documents the decision, e.g.:

```suggestion
ఇద్దరు వేల~౨౦౦౦
```

**Review by folivoramanh — COMMENTED** [review]
New language for the repo (Telugu ITN, Cardinal class). The lakh/crore tier architecture (fixed-width digit-block concatenation + leading-zero strip, structural `at_least_one_non_zero` disambiguation for the crore recursion, and the separate "NN వందలు" hundred-multiplier idiom) is sound — 67/67 pytest pass and every tier boundary I derived and ran (lakh boundary, crore boundary, 14-nines ceiling) produced correct output. One open question below on `digit_people` scope, otherwise just nits.

File: tests/nemo_text_processing/te/test_sparrowhawk_inverse_text_normalization.sh [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+PROJECT_DIR=/workspace/tests
+
+runtest () {
+  input=$1
+  cd /workspace/sparrowhawk/documentation/grammars
+
+  while read testcase; do
+    IFS='~' read spoken written <<< $testcase
+    denorm_pred=$(echo $spoken | normalizer_main --config=sparrowhawk_configuration.ascii_proto 2>&1 | tail -n 1)
```
Comment (folivoramanh) [inline tests/nemo_text_processing/te/test_sparrowhawk_inverse_text_normalization.sh:11]: **optional** · derived

Unlike sibling scripts for other languages (e.g. `kn`'s, which escapes backslashes and folds `\xC2\xA0` non-breaking spaces back to regular spaces before comparing), this one does neither. Low risk given the current test data has no backslashes, but worth aligning with the established template if the Telugu test set grows to include Sparrowhawk edge cases later.

File: nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py (hunk 17 above)
Comment (jaya-TN) [reply nemo_text_processing/inverse_text_normalization/te/taggers/cardinal.py]: Scoped graph_people to standalone person-count usage so person-count forms are no longer treated as generic magnitude multipliers. Standard numeric forms remain unchanged, and all 67 Cardinal unit tests pass.

File: tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt (hunk 18 above)
Comment (jaya-TN) [reply tests/nemo_text_processing/te/data_inverse_text_normalization/test_cases_cardinal.txt:67]: Scoped graph_people to standalone person-count usage, so person-count forms are no longer treated as magnitude multipliers. Therefore, the suggested ఇద్దరు వేల → ౨౦౦౦ case is not applicable.

File: tests/nemo_text_processing/te/test_sparrowhawk_inverse_text_normalization.sh (hunk 19 above)
Comment (jaya-TN) [reply tests/nemo_text_processing/te/test_sparrowhawk_inverse_text_normalization.sh:11]: Telugu ITN follows the existing Hindi ITN Sparrowhawk template (spoken~written parsing and whitespace trimming). The current Telugu test data has no backslash or non-breaking-space cases, so I’m keeping the current implementation unchanged.

**Review by folivoramanh — APPROVED** [review]
